### PR TITLE
Add new methods to WPAuthenticatorDelegate to allow host apps to handle errors

### LINKED ISSUE #530: Allow host apps to handle errors
At the moment most of the most common errors (i.e. "the site is not a WP site") are handled by WPAuthenticator.

Host apps might want to take over some of those errors. In fact, WooCommerce wants to be able to know about certain errors and want to provide custom ui for them. (see, for reference, woocommerce/woocommerce-ios#3218)

### PR DESCRIPTION
**Pull request by ctarda**
Closes #530 
Ref: p91TBi-3Rn-p2

## Changes
* Added two new methods to `WPAuthenticatorDelegate` to allow host apps to decide if they want to handle errors.
* Bumped the podspec
* Extracted a private method in SiteAddressViewController to execute the insertion of custom UI

## How to test
* This PR provides the functionality for woocommerce/woocommerce-ios#3219 so the best way to test is probably by following the test steps in the linked PR to WooCommerce.
* Please inspect the diff. Let me know if there is a better way 😊

### CODE REVIEW

File: WordPressAuthenticator/Authenticator/WordPressAuthenticatorDelegateProtocol.swift
Comment (ctarda) [inline WordPressAuthenticator/Authenticator/WordPressAuthenticatorDelegateProtocol.swift:71]: These two methods would be new. This would break WordPress, but the fix is simple: for host apps that are good with WPAuthenticator's error handling, shoudlDisplayError would just return false, and handleError would do nothing

File: WordPressAuthenticator/Unified Auth/View Related/Site Address/SiteAddressViewController.swift
Comment (ctarda) [inline WordPressAuthenticator/Unified Auth/View Related/Site Address/SiteAddressViewController.swift:143]: This might not be necessary. In fact, I would argue that it would be good to avoid doing it.

File: WordPressAuthenticator/Unified Auth/View Related/Site Address/SiteAddressViewController.swift
Comment (ctarda) [inline WordPressAuthenticator/Unified Auth/View Related/Site Address/SiteAddressViewController.swift]: This is what actually inserts the custom UI for the case when the url provided for login with site address is not a WordPress site. This code would use some cleanup, but I hope it helps illustrate how this solution would work

File: WordPressAuthenticator/Authenticator/WordPressAuthenticatorDelegateProtocol.swift
Comment (ctarda) [inline WordPressAuthenticator/Authenticator/WordPressAuthenticatorDelegateProtocol.swift]: The parameter in the completion block might just be of type UIViewController. After all, host apps would only call this method if they actually want to inject a custom error UI

File: WordPressAuthenticator/Authenticator/WordPressAuthenticatorDelegateProtocol.swift
Comment (ScoutHarris) [reply WordPressAuthenticator/Authenticator/WordPressAuthenticatorDelegateProtocol.swift:71]: Thanks for the info! After this is merged, I'll create a WP PR to address this.

File: WordPressAuthenticator/Unified Auth/View Related/Site Address/SiteAddressViewController.swift
Comment (ScoutHarris) [reply WordPressAuthenticator/Unified Auth/View Related/Site Address/SiteAddressViewController.swift:143]: I'm sorry, but I'm confused by this. It appears you added this override method to implement your error handling. Why do you not agree with what you just did? 😄

File: WordPressAuthenticator/Unified Auth/View Related/Site Address/SiteAddressViewController.swift
Comment (ScoutHarris) [inline WordPressAuthenticator/Unified Auth/View Related/Site Address/SiteAddressViewController.swift]: Can this just be `guard authenticationDelegate.shouldHandleError(error) else {`?

**Review by mindgraffiti — APPROVED** [review]
This approach seems reasonable to me, though I am curious about your answer to Stephenie's question (about why you don't like overriding the displayRemoteError) 😀

File: WordPressAuthenticator/Unified Auth/View Related/Site Address/SiteAddressViewController.swift
Comment (mindgraffiti) [reply WordPressAuthenticator/Unified Auth/View Related/Site Address/SiteAddressViewController.swift]: iirc, this is part of Woo's coding style. I could be wrong though.

File: WordPressAuthenticator/Unified Auth/View Related/Site Address/SiteAddressViewController.swift
Comment (ctarda) [reply WordPressAuthenticator/Unified Auth/View Related/Site Address/SiteAddressViewController.swift:143]: Sorry, that was terrible wording on my side. Let me rephrase my comment:

"I am not 100% sure that this is absolutely necessary. So far I have not seen this override being used in any of the tests I have run using WooCommerce. That does not mean it won't happen as I continue testing. I would prefer if, eventually, we noticed that overriding this method is actually unnecessary and we could get rid of the override"

File: WordPressAuthenticator/Unified Auth/View Related/Site Address/SiteAddressViewController.swift
Comment (ctarda) [reply WordPressAuthenticator/Unified Auth/View Related/Site Address/SiteAddressViewController.swift]: Yes, it's probably muscle memory. I'll update it

File: WordPressAuthenticator/Unified Auth/View Related/Site Address/SiteAddressViewController.swift
Comment (ctarda) [reply WordPressAuthenticator/Unified Auth/View Related/Site Address/SiteAddressViewController.swift:143]: And I just found a test case for WooCommerce where these overrides are necessary: a Pressable store with Jetpack installed but not activated

**Comment by ctarda** [comment]
Thanks for the review @mindgraffiti @ScoutHarris 

I bumped the podspec, updated the guard clause and hopefully clarified my comment about the override. 

I'll merge when CI is ✅

File: WordPressAuthenticator/Unified Auth/View Related/Site Address/SiteAddressViewController.swift
Comment (ScoutHarris) [reply WordPressAuthenticator/Unified Auth/View Related/Site Address/SiteAddressViewController.swift:143]: Oh! OK, that makes more sense. 😄 Thank you sir!

**Pull request merged**
